### DISCUSSION TITLE
How to use user_id of logged-in user to filter data?

### DISCUSSION THREAD

**Discussion by novotny1akub** [observablehq/framework#1099] [Q&A] [answered]
This is related to https://github.com/observablehq/framework/issues/1039 (selective access to certain rows based on the user that is logged in). I tried to come up with own solution using `Express` web framework. In my POC, I have a manually created `login.html` file that dispatches a `POST` request. The processing of this request occurs in `app.js`, where I implement session authentication. Upon successful authentication, the user gains access to all HTML files within a designated folder specified in `app.js`.  The idea is to use this setup to expose the `dist` folder created by `Observable Framework` (created using `npm run build`). This all works fine as far as I can tell. However, I am not sure how to pass the information about the logged-in user id and subsequently how to use the id to filter data with Observable.
Some online suggestions involve using templating tools like EJS, but those are typically for EJS files, not Markdown like what I am dealing with (something like `ejs.renderFile('file.ejs', { username }, (err, html) => { res.send(html); })` ). I tried to use ejs.renderFile for html files generated by Observable, but it does not seem to be working. I ventured into Node.js primarily because of Observable, and it has only been a week since I started delving into it so there is a chance I have goofed up somewhere.

So, my question is twofold: Is this approach a bad idea? And secondly, is it possible to send data from Express middleware for use in calculations within Observable using `EJS` or some other tool?

**Comment by Fil** [comment]
Do you really need to send a personalized "page" (html), or would it be enough to send a personalized "data" (say, csv file contents)? The first approach seems a bit more difficult to me, because the markdown page has to go through Framework to be converted to HTML and sent to the browser, and this happens when we build the site. The second approach might be easier, since you'd only personalize one "raw" file (or several).

(Pardon me if I don't fully grok the scope of your project, it seems to be a complex construction.)

**Reply by novotny1akub** [reply]
I only need the personalised data (sorry, put it in another thread). I have been thinking about your answer and it appears that fetching the data dynamically without relying on data loaders would be the most suitable approach. However, it seems that the Observable Framework documentation does not address this method. My main challenge is incorporating dynamically fetched data into markdown and/or utilizing `<div>${dynamically_loaded_data}</div>`.

**Reply by novotny1akub** [reply, marked as answer]
I must have messed up something because when I tried a more simplified example today, it works without any problems. Apologies for the question. Here is my example of dynamically loaded data:

````
---
toc: false
---

# Fetching Data Dynamically

Select a Star Wars character id using the slider (https://swapi.dev/documentation)
```js
const star_wars_id = view(Inputs.range([1, 15], {step: 1, value: 4}));
```
<div>
  <b>id=${star_wars_id}</b>
  <p>Star Wars character name: <b>${fetchData()}</b> <p> 
  <p>(character name loaded dynamically from an API)</p>
</div>

```js

async function fetchData(){

    const url = `https://akabab.github.io/starwars-api/api/id/${star_wars_id}.json`;
    const response = await fetch(url);
    const data = await response.json();

    return data.name
}

```
````
![image](https://github.com/observablehq/framework/assets/39886440/7a22bf53-573e-4a39-8a96-112f8f157a83)
